lib/ecto/parameterized_type.ex

```diff
@@ -0,0 +1,82 @@
+defmodule Ecto.ParameterizedType do
+  @moduledoc """
+  Parameterized types are similar to Ecto Custom Types, however they allow a set of options to be
+  specified in the schema which are initialized on compilation and passed to the callback functions
+  as a final argument. Note that options should be specified as a keyword, but it is idiomatic to
+  convert them to maps inside init for easier pattern matching in other callbacks.
+
+  Note that as opposed to regular types, custom types are responsible for handling nil in both load
+  and dump callbacks.
+
+  ## Examples
+
+    To create a parameterized type, create a module as shown below:
+
+    defmodule MyApp.MyType do
+      use Ecto.ParameterizedType
+
+      def type(_params), do: :string
+
+      def init(opts) do
+        validate_opts(opts)
+        Enum.into(opts, %{})
+      end
+
+      def cast(data, params) do
+        ...
+        cast_data
+      end
+
+      def load(data, _loader, params) do
+        ...
+        {:ok, loaded_data}
+      end
+
+      def dump(data, dumper, params) do
+        ...
+        {:ok, dumped_data}
+      end
+
+      def equal?(a, b, _params) do
+        a == b
+      end
+    end
+
+  To use this type in a schema field, specify the type and parameters like this:
+
+    schema "foo" do
+      field "bar", MyApp.MyType, opt1: :baz, opt2: :boo
+    end
+
+  """
+
+  @type opts :: keyword()
+
+  @type params :: term()
+
+  @callback init(opts :: opts()) :: params()
+
+  @callback cast(data :: term, params :: params()) ::
+              {:ok, term} | {:error, keyword()} | :error
+
+  @callback load(value :: any(), loader :: function(), params :: params()) :: {:ok, value :: any()} | :error
+
+  @callback dump(value :: any(), dumper :: function(), params :: params()) :: {:ok, value :: any()} | :error
+
+  @callback type(params :: params()) :: Ecto.Type.t()
+
+  @callback equal?(value1 :: any(), value2 :: any(), params :: params()) :: boolean()
+
+  @callback embed_as(format :: atom(), params :: params()) :: :self | :dump
+
+  @doc false
+  defmacro __using__(_) do
+    quote location: :keep do
+      @behaviour Ecto.ParameterizedType
+      def embed_as(_, _), do: :self
+      def equal?(term1, term2, _params), do: term1 == term2
+      def init(opts), do: Enum.into(opts, %{})
```

Comment:
Let's remove the default implementation for `init`. I don't think it is a good idea for the default to simply collecting everything. Typically, you will want to validate options and keep only what you need. You will also want to precompute as much as you can to avoid compile time work. For example, for enums, you would do something like this:

```elixir
def init(opts) do
  values = Keyword.fetch!(opts, :values)
  user_to_db = Map.new(values, fn k -> {k, Atom.to_string(v)} end)
  db_to_user = Map.new(values, fn k -> {Atom.to_string(v), k} end)
  %{user_to_db: user_to_db, db_to_user: db_to_user}
end
```

So load and dump happens as efficiently as possible. For `Ecto.Embedded`, you can make it so:

```elixir
def init(_opts), do: raise "TODO: invoke this directly"
```

Reply:
Done. I really like this mapping method, it makes cast/load/dump much nicer as declarative instead of imperative.